README
NAME
    Net::API::RPX - Perl interface to Janrain's RPX service

VERSION
    version 0.04

SYNOPSIS
        use Net::API::RPX;

        my $rpx = Net::API::RPX->new({ api_key => '<your_api_key_here>' });

        $rpx->auth_info({ token => $token });

DESCRIPTION
    This module is a simple wrapper around Janrain's RPX service. RPX
    provides a single method for dealing with third-party authentication.

    See <http://www.rpxnow.com> for more details.

    For specific information regarding the RPX API and method arguments,
    please refer to <https://rpxnow.com/docs>.

ATTRIBUTES
    This is a Moose based module, this classes attribtues are as so:

  api_key
    This is the api_key provided by Janrain to interface with RPX. You will
    need to signup to RPX to get one of these.

  base_url
    This is the base URL that is used to make API calls against. It defaults
    to the RPX v2 API.

  ua
    This is a LWP::UserAgent object. You may override it if you require more
    fine grain control over remote queries.

METHODS
  auth_info
        my $user_data = $rpx->auth_info({ token => $params{token} });

    Upon redirection back from RPX, you will be supplied a token to use for
    verification. Call auth_info to verify the authenticity of the token and
    gain user details.

    'token' argument is required, 'extended' argument is optional.

  map
        $rpx->map({ identifier => 'yet.another.open.id', primary_key => 12 });

    This method allows you to map more than one 'identifier' to a user.

    'identifier' argument is required, 'primary_key' argument is required,
    'overwrite' is optional.

  unmap
        $rpx->unmap({ identifier => 'yet.another.open.id', primary_key => 12 });

    This is the inverse of 'map'.

    'identifier' argument is required, 'primary_key' argument is required.

  mappings
        my $data = $rpx->mappings({ primary_key => 12 });

    This method returns information about the identifiers associated with a
    user.

    'primary_key' argument is required.

TEST COVERAGE
    This distribution is heavily unit and system tested for compatability
    with Test::Builder. If you come across any bugs, please send me or
    submit failing tests to Net-API-RPX RT queue. Please see the 'SUPPORT'
    section below on how to supply these.

     ---------------------------- ------ ------ ------ ------ ------ ------ ------
     File                           stmt   bran   cond    sub    pod   time  total
     ---------------------------- ------ ------ ------ ------ ------ ------ ------
     blib/lib/Net/API/RPX.pm       100.0  100.0    n/a  100.0  100.0  100.0  100.0
     Total                         100.0  100.0    n/a  100.0  100.0  100.0  100.0
     ---------------------------- ------ ------ ------ ------ ------ ------ ------

BUGS
    Please report any bugs or feature requests to "bug-net-api-rpx at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-API-RPX>. I will be
    notified, and then you'll automatically be notified of progress on your
    bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Net::API::RPX

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-API-RPX>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Net-API-RPX>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Net-API-RPX>

    *   Search CPAN

        <http://search.cpan.org/dist/Net-API-RPX>

SEE ALSO
    <http://www.janrain.com/>, <http://www.rpxnow.com/>

AUTHORS
    *   Scott McWhirter <konobi@cpan.org>

    *   Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2012 by Cloudtone Studios.

    This is free software, licensed under:

      The (three-clause) BSD License



Changes
Revision history for Net-API-RPX

0.04 2012-05-24T00:27:53Z

 - Re-released to avoid a lot of cruft that snuck in somehow ( WTF?!!? )

0.03 2012-05-24T00:23:24Z

 - Toolkit switched to Dist::Zilla ( Kent Fredric )

0.02 2012-05-23T23:31:00Z
 - Added fully fledged exception objects to replace die() calls.
 - If you catch one somehow, you can now introspect them and get a better
   idea of what went wrong. ( Kent Fredric )

0.01 2009-02-20T12:14:00Z
 - First version.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Data::Dump" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "mocked" => 0,
    "unmocked" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Perl interface to Janrain's RPX service",
  "dist_author" => [
    "Scott McWhirter <konobi\@cpan.org>",
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Net-API-RPX",
  "dist_version" => "0.04",
  "license" => "bsd",
  "module_name" => "Net::API::RPX",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "JSON::Any" => 0,
    "LWP::UserAgent" => 0,
    "Moose" => 0,
    "Throwable::Error" => 0,
    "Throwable::X" => 0,
    "URI" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Net/API/RPX/Exception.pm
use strict;
package Net::API::RPX::Exception;
BEGIN {
  $Net::API::RPX::Exception::AUTHORITY = 'cpan:KONOBI';
}
{
  $Net::API::RPX::Exception::VERSION = '0.04';
}

# ABSTRACT:  A Baseclass for Net::API::RPX exceptions based on L<Throwable::X>

use warnings;

use Moose;

extends 'Throwable::Error';

with qw( Throwable::X );

use Throwable::X -all;

sub _immutable {
    (shift)->meta->make_immutable( inline_constructor => 0 );
}
__PACKAGE__->_immutable;

1;


__END__
=pod

=head1 NAME

Net::API::RPX::Exception - A Baseclass for Net::API::RPX exceptions based on L<Throwable::X>

=head1 VERSION

version 0.04

=head1 AUTHORS

=over 4

=item *

Scott McWhirter <konobi@cpan.org>

=item *

Kent Fredric <kentnl@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Cloudtone Studios.

This is free software, licensed under:

  The (three-clause) BSD License

=cut



lib/Net/API/RPX/Exception/Usage.pm
use strict;
package Net::API::RPX::Exception::Usage;
BEGIN {
  $Net::API::RPX::Exception::Usage::AUTHORITY = 'cpan:KONOBI';
}
{
  $Net::API::RPX::Exception::Usage::VERSION = '0.04';
}

# ABSTRACT: For when the interface is used wrongly
use warnings;

use Moose;
use namespace::autoclean;

extends 'Net::API::RPX::Exception';

has 'required_parameter' => ( isa => 'Str', is => 'rw', required => 1 );
has 'method_name'        => ( isa => 'Str', is => 'rw', required => 1 );
has 'signature'          => ( isa => 'Str', is => 'rw', required => 1 );
has 'package'            => ( isa => 'Str', is => 'rw', required => 1 );

sub _signature_string {
    my ($self) = @_;
    return $self->method_name . '(' . $self->signature . ')';
}

sub _explanation {
    my ($self) = @_;
    return
        sprintf q{Method %s on package %s expects the parameter "%s"} 
      . qq{\n\n}
      . qq{\tUsage:\t%s\n}, $self->method_name,
      $self->package,
      $self->required_parameter, $self->_signature_string;
}

around message => sub {
    my ( $orig, $self, @args ) = @_;
    return $self->$orig(@args) . qq{\n} . $self->_explanation;
};

__PACKAGE__->_immutable;
1;


__END__
=pod

=head1 NAME

Net::API::RPX::Exception::Usage - For when the interface is used wrongly

=head1 VERSION

version 0.04

=head1 AUTHORS

=over 4

=item *

Scott McWhirter <konobi@cpan.org>

=item *

Kent Fredric <kentnl@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Cloudtone Studios.

This is free software, licensed under:

  The (three-clause) BSD License

=cut



lib/Net/API/RPX/Exception/Network.pm
use strict;
package Net::API::RPX::Exception::Network;
BEGIN {
  $Net::API::RPX::Exception::Network::AUTHORITY = 'cpan:KONOBI';
}
{
  $Net::API::RPX::Exception::Network::VERSION = '0.04';
}

# ABSTRACT: A Class of exceptions for network connectivitiy issues.

use warnings;
use Moose;
use namespace::autoclean;

extends 'Net::API::RPX::Exception';

has 'ua_result'   => ( isa => "Ref", is => 'ro', required => 1 );
has 'status_line' => ( isa => "Str", is => 'ro', required => 1 );

__PACKAGE__->_immutable;
1;


__END__
=pod

=head1 NAME

Net::API::RPX::Exception::Network - A Class of exceptions for network connectivitiy issues.

=head1 VERSION

version 0.04

=head1 AUTHORS

=over 4

=item *

Scott McWhirter <konobi@cpan.org>

=item *

Kent Fredric <kentnl@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Cloudtone Studios.

This is free software, licensed under:

  The (three-clause) BSD License

=cut



lib/Net/API/RPX/Exception/Service.pm
use strict;
package Net::API::RPX::Exception::Service;
BEGIN {
  $Net::API::RPX::Exception::Service::AUTHORITY = 'cpan:KONOBI';
}
{
  $Net::API::RPX::Exception::Service::VERSION = '0.04';
}

# ABSTRACT: A Class of exceptions for delivering problems from the RPX service.

use warnings;
use Moose;
use namespace::autoclean;

extends 'Net::API::RPX::Exception';


my $rpx_errors = {
    -1 => 'Service Temporarily Unavailable',
    0  => 'Missing parameter',
    1  => 'Invalid parameter',
    2  => 'Data not found',
    3  => 'Authentication error',
    4  => 'Facebook Error',
    5  => 'Mapping exists',
};

has 'data'              => ( isa => 'Any', is => 'ro', required => 1 );
has 'status'            => ( isa => 'Any', is => 'ro', required => 1 );
has 'rpx_error'         => ( isa => 'Any', is => 'ro', required => 1 );
has 'rpx_error_code'    => ( isa => 'Any', is => 'ro', required => 1 );
has 'rpx_error_message' => ( isa => 'Any', is => 'ro', required => 1 );
has
  'rpx_error_code_description' =>
  ( isa => 'Any', is => 'ro', required => 1, lazy => 1, ),
  , builder => '_build_rpx_error_code_description';

sub _build_rpx_error_code_description {
    my ($self) = shift;
    return $rpx_errors->{ $self->rpx_error_code };
}
__PACKAGE__->_immutable();
1;


__END__
=pod

=head1 NAME

Net::API::RPX::Exception::Service - A Class of exceptions for delivering problems from the RPX service.

=head1 VERSION

version 0.04

=head1 NAME

=head1 AUTHORS

=over 4

=item *

Scott McWhirter <konobi@cpan.org>

=item *

Kent Fredric <kentnl@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Cloudtone Studios.

This is free software, licensed under:

  The (three-clause) BSD License

=cut



lib/Net/API/RPX.pm
use strict;
package Net::API::RPX;
BEGIN {
  $Net::API::RPX::AUTHORITY = 'cpan:KONOBI';
}
{
  $Net::API::RPX::VERSION = '0.04';
}

# ABSTRACT: Perl interface to Janrain's RPX service

use Moose;
use LWP::UserAgent;
use URI;
use JSON::Any;
use Net::API::RPX::Exception::Usage;
use Net::API::RPX::Exception::Network;
use Net::API::RPX::Exception::Service;

has api_key => (
    is       => 'rw',
    isa      => 'Str',
    required => 1,
);

has base_url => (
    is       => 'rw',
    isa      => 'Str',
    required => 1,
    lazy     => 1,
    default  => 'https://rpxnow.com/api/v2/',
);

has ua => (
    is       => 'rw',
    isa      => 'Object',
    required => 1,
    lazy     => 1,
    builder  => '_build_ua',
);

sub _build_ua {
    my ($self) = @_;
    return LWP::UserAgent->new( agent => $self->_agent_string );
}

has _agent_string => (
    is       => 'rw',
    isa      => 'Str',
    required => 1,
    lazy     => 1,
    default  => sub { 'net-api-rpx-perl/' . $Net::API::RPX::VERSION },
);


sub auth_info {
    my ( $self, $opts ) = @_;
    Net::API::RPX::Exception::Usage->throw(
        ident              => 'auth_info_usage_needs_token',
        message            => "Token is required",
        required_parameter => 'token',
        method_name        => '->auth_info',
        package            => __PACKAGE__,
        signature          => '{ token => $authtoken }',
    ) if !exists $opts->{token};
    return $self->_fetch( 'auth_info', $opts );
}


sub map {
    my ( $self, $opts ) = @_;
    Net::API::RPX::Exception::Usage->throw(
        ident              => 'map_usage_needs_identifier',
        message            => "Identifier is required",
        required_parameter => 'identifier',
        method_name        => '->map',
        package            => __PACKAGE__,
        signature => '{ identifier => \'some.open.id\', primary_key => 12 }',
    ) if !exists $opts->{identifier};

    Net::API::RPX::Exception::Usage->throw(
        ident              => 'map_usage_needs_primary_key',
        message            => "Primary Key is required",
        required_parameter => 'primary_key',
        method_name        => '->map',
        package            => __PACKAGE__,
        signature => '{ identifier => \'some.open.id\', primary_key => 12 }',
    ) if !exists $opts->{primary_key};
    $opts->{primaryKey} = delete $opts->{primary_key};

    return $self->_fetch( 'map', $opts );
}


sub unmap {
    my ( $self, $opts ) = @_;
    Net::API::RPX::Exception::Usage->throw(
        ident              => 'unmap_usage_needs_identifier',
        message            => "Identifier is required",
        required_parameter => 'identifier',
        method_name        => '->unmap',
        package            => __PACKAGE__,
        signature => '{ identifier => \'some.open.id\', primary_key => 12 }',
    ) if !exists $opts->{identifier};

    Net::API::RPX::Exception::Usage->throw(
        ident              => 'unmap_usage_needs_primay_key',
        message            => "Primary Key is required",
        required_parameter => 'primary_key',
        method_name        => '->unmap',
        package            => __PACKAGE__,
        signature => '{ identifier => \'some.open.id\', primary_key => 12 }',
    ) if !exists $opts->{primary_key};

    $opts->{primaryKey} = delete $opts->{primary_key};

    return $self->_fetch( 'unmap', $opts );
}


sub mappings {
    my ( $self, $opts ) = @_;
    Net::API::RPX::Exception::Usage->throw(
        ident              => 'mappings_usage_needs_primary_key',
        message            => "Primary Key is required",
        required_parameter => 'primary_key',
        method_name        => '->mappings',
        package            => __PACKAGE__,
        signature          => '{ primary_key => 12 }',
    ) if !exists $opts->{primary_key};

    $opts->{primaryKey} = delete $opts->{primary_key};

    return $self->_fetch( 'mappings', $opts );
}

my $rpx_errors = {
    -1 => 'Service Temporarily Unavailable',
    0  => 'Missing parameter',
    1  => 'Invalid parameter',
    2  => 'Data not found',
    3  => 'Authentication error',
    4  => 'Facebook Error',
    5  => 'Mapping exists',
};

sub _fetch {
    my ( $self, $uri_part, $opts ) = @_;

    my $uri = URI->new( $self->base_url . $uri_part );
    my $res = $self->ua->post(
        $uri,
        {
            %$opts,
            apiKey => $self->api_key,
            format => 'json',
        }
    );

    if ( !$res->is_success ) {
        Net::API::RPX::Exception::Network->throw(
            ident       => '_fetch_network_failure',
            message     => "Could not contact RPX: " . $res->status_line(),
            ua_result   => $res,
            status_line => $res->status_line,
        );
    }

    my $data = JSON::Any->from_json( $res->content );
    if ( $data->{'stat'} ne 'ok' ) {
        my $err = $data->{'err'};
        Net::API::RPX::Exception::Service->throw(
            ident             => '_fetch_service_error',
            data              => $data,
            status            => $data->{'stat'},
            rpx_error         => $data->{'err'},
            rpx_error_code    => $data->{err}->{code},
            rpx_error_message => $data->{err}->{msg},
            message           => "RPX returned error of type '"
              . $rpx_errors->{ $err->{code} }
              . "' with message: "
              . $err->{msg},
        );
    }
    delete $data->{'stat'};
    return $data;
}

1;    # End of Net::API::RPX

__END__
=pod

=head1 NAME

Net::API::RPX - Perl interface to Janrain's RPX service

=head1 VERSION

version 0.04

=head1 SYNOPSIS

    use Net::API::RPX;

    my $rpx = Net::API::RPX->new({ api_key => '<your_api_key_here>' });

    $rpx->auth_info({ token => $token });

=head1 DESCRIPTION

This module is a simple wrapper around Janrain's RPX service. RPX provides a single method for
dealing with third-party authentication.

See L<http://www.rpxnow.com> for more details.

For specific information regarding the RPX API and method arguments, please refer to
L<https://rpxnow.com/docs>.

=head1 ATTRIBUTES

This is a Moose based module, this classes attribtues are as so:

=head2 api_key

This is the api_key provided by Janrain to interface with RPX. You will need to signup to RPX
to get one of these.

=head2 base_url

This is the base URL that is used to make API calls against. It defaults to the RPX v2 API.

=head2 ua

This is a LWP::UserAgent object. You may override it if you require more fine grain control
over remote queries.

=head1 METHODS

=head2 auth_info

    my $user_data = $rpx->auth_info({ token => $params{token} });

Upon redirection back from RPX, you will be supplied a token to use for verification. Call
auth_info to verify the authenticity of the token and gain user details.

'token' argument is required, 'extended' argument is optional.

=head2 map

    $rpx->map({ identifier => 'yet.another.open.id', primary_key => 12 });

This method allows you to map more than one 'identifier' to a user.

'identifier' argument is required, 'primary_key' argument is required, 'overwrite' is optional.

=head2 unmap

    $rpx->unmap({ identifier => 'yet.another.open.id', primary_key => 12 });

This is the inverse of 'map'.

'identifier' argument is required, 'primary_key' argument is required.

=head2 mappings

    my $data = $rpx->mappings({ primary_key => 12 });

This method returns information about the identifiers associated with a user.

'primary_key' argument is required.

=head1 TEST COVERAGE

This distribution is heavily unit and system tested for compatability with
L<Test::Builder>. If you come across any bugs, please send me or submit failing
tests to Net-API-RPX RT queue. Please see the 'SUPPORT' section below on
how to supply these.

 ---------------------------- ------ ------ ------ ------ ------ ------ ------
 File                           stmt   bran   cond    sub    pod   time  total
 ---------------------------- ------ ------ ------ ------ ------ ------ ------
 blib/lib/Net/API/RPX.pm       100.0  100.0    n/a  100.0  100.0  100.0  100.0
 Total                         100.0  100.0    n/a  100.0  100.0  100.0  100.0
 ---------------------------- ------ ------ ------ ------ ------ ------ ------

=head1 BUGS

Please report any bugs or feature requests to C<bug-net-api-rpx at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-API-RPX>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Net::API::RPX

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-API-RPX>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Net-API-RPX>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Net-API-RPX>

=item * Search CPAN

L<http://search.cpan.org/dist/Net-API-RPX>

=back

=head1 SEE ALSO

L<http://www.janrain.com/>, L<http://www.rpxnow.com/>

=head1 AUTHORS

=over 4

=item *

Scott McWhirter <konobi@cpan.org>

=item *

Kent Fredric <kentnl@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Cloudtone Studios.

This is free software, licensed under:

  The (three-clause) BSD License

=cut

